src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexNode.cs

```diff
@@ -1064,304 +907,250 @@ internal TransitionRegex<S> CreateDerivative()
+                            return RightTransition(this, elem, context, transitions, continuation, effects, eager);
+                        }
                     }
-                    break;
-
-                case SymbolicRegexNodeKind.Not:
-                    Debug.Assert(_left is not null);
-                    transition = _left.CreateDerivativeWithEffects(eager).Complement();
-                    break;
-
-                default:
-                    transition = TransitionRegex<S>.Leaf(_builder._nothing);
-                    break;
-            }
-            return transition;
-        }
-
-        /// <summary>
-        /// Wrap a TransitionRegex with the effects under this node. The effects are valid when this node is nullable.
-        /// </summary>
-        /// <remarks>
-        /// The construction follows the paths that the backtracking matcher would take. For example in ()|() only the effects for the first
-        /// alternative will be included.
-        /// </remarks>
-        internal TransitionRegex<S> WrapEffects(TransitionRegex<S> transition)
-        {
-            if (!StackHelper.TryEnsureSufficientExecutionStack())
-            {
-                return StackHelper.CallOnEmptyStack(WrapEffects, transition);
-            }
-
-            switch (_kind)
-            {
-                case SymbolicRegexNodeKind.Concat:
-                    Debug.Assert(_left is not null && _right is not null);
-                    Debug.Assert(_left.CanBeNullable && _right.CanBeNullable);
-                    transition = _left.WrapEffects(transition);
-                    transition = _right.WrapEffects(transition);
-                    break;
 
                 case SymbolicRegexNodeKind.Loop:
                     Debug.Assert(_left is not null);
-                    // Apply effect when backtracking engine would enter loop
-                    if (_lower != 0)
-                    {
-                        Debug.Assert(_left.CanBeNullable);
-                        transition = _left.WrapEffects(transition);
-                    }
-                    else if (_upper != 0 && !IsLazy && _left.CanBeNullable)
+                    Debug.Assert(_upper > 0);
+
+                    // Add transitions only when the backtracking engines would prefer to enter the loop
+                    if (!eager || !IsLazy || _lower != 0)
                     {
-                        TransitionRegex<S> bodyTransition = _left.WrapEffects(transition);
-                        transition = _left.IsNullable ?
-                            bodyTransition :
-                            TransitionRegex<S>.Lookaround(_left.ExtractNullabilityTest(),
-                                bodyTransition,
-                                transition);
+                        // The loop derivative peels out one iteration and concatenates the body's derivative with the decremented loop,
+                        // so if n,m > 0 then d(R{n,m}) = d(R)R{n-1,m-1}.
```

Comment:
What happens if the lower bound is already 0?

Also, total nit, but RegexNode.M is used as the lower bound and RegexNode.N is the upper bound, but this comment is using n as a lower bound and m as an upper.  It's not a big deal, but if we could consistently use m for lower and n for upper, that'd be helpful.